### DISCUSSION TITLE
Wasmer's metering middleware for contracts

### DISCUSSION THREAD

**Discussion by nazar-pc** [paritytech/substrate#11143] [Ideas]
I see there is work ongoing in https://github.com/paritytech/substrate/issues/10298 to make wasmer singlepass a viable option.

I have not found any discussions on wasmer's metering feature though: https://docs.rs/wasmer-middleware-common/0.17.1/wasmer_middleware_common/metering/struct.Metering.html

It seems like it should be a substantial improvement due to avoiding host calls in the process of metering execution.

I found it from this useful NEAR discussion: https://gov.near.org/t/on-running-some-wasm-contracts-with-llvm-specifically-for-evm/2869/4

Just wanted to make sure those working on pallet-contracts are aware of it and maybe get some feedback in case it was already considered and rejected for some reason.

**Comment by athei** [comment]
It is not as easy as it seems. Depending on an execution engine specific implementation of gas metering is a dangerous thing. You will tie your consensus to an implementation detail. You will not be able to run contracts with different execution engines.

It is still totally unclear if singlepass compilation brings benefits for contracts when considering compilation time. We are trying to delay this decision as much as possible. This is why injected gas metering is what we do for now.

**Reply by nazar-pc** [reply]
I recommend reading NEAR discussion link in more detail if you have time, specifically here (so I don't have to quote it): https://gov.near.org/t/on-running-some-wasm-contracts-with-llvm-specifically-for-evm/2869/9

TL;DR: It is not VM-specific, in fact NEAR folks wrote VM-independent instrumentation themselves based on that approach as far as I can see (before that they used `pwasm-utils` too).

**Reply by athei** [reply]
wasmer just injects a global variable for gas counting instead of just calling into the host. We can implement that in `wasm-instrument`. I believe NEAR did just that. The post you linked is old. We just didn't came around to do this performance optimization.

**Reply by nazar-pc** [reply]
Awesome, yeah, that is all I wanted to know, thanks!
